### PR TITLE
draft: distributed data parallel with pytorch native

### PR DESCRIPTION
**Pull request by azhou-determined**
## Description

- Get rid of hvd_config
- Test model for native pytorch DDP
## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: examples/tutorials/mnist_pytorch_dtrain/README.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# PyTorch MNIST CNN Tutorial
```
Comment (rb-determined-ai) [inline examples/tutorials/mnist_pytorch_dtrain/README.md:1]: is the example just copy/pasted, for testing purposes?

File: harness/determined/_execution.py [hunk 2]
```diff
@@ -128,11 +128,11 @@ def _make_local_execution_env(
         test_mode=test_mode,
         on_cluster=False,
     )
-    hvd_config = horovod.HorovodContext.from_configs(env.experiment_config, env.hparams)
+    # hvd_config = horovod.HorovodContext.from_configs(env.experiment_config, env.hparams)
```
Comment (rb-determined-ai) [inline harness/determined/_execution.py:131]: don't forget to delete this comment

File: harness/determined/_generic/_distributed.py [hunk 3]
```diff
@@ -34,6 +35,7 @@ def __init__(
         local_size: int,
         cross_rank: int,
         cross_size: int,
+            backend: Optional[str],
```
Comment (rb-determined-ai) [inline harness/determined/_generic/_distributed.py:38]: I don't think this is the right way to pass the backend information around.  The backend is 'training loop-specific information', while the DistributedContext is more like a 'training-loop-independent way to ensure a minimum set of distributed communication primitives'

In the case of a totally custom training loop, the backend would just be some random string that the user sets and knows how to interpret, and that's a strange thing to attach to this object I think.

File: harness/determined/_trial_controller.py [hunk 4]
```diff
@@ -38,20 +36,17 @@ def __init__(
...
         self.scheduling_unit = self.env.experiment_config.scheduling_unit()
 
-        if self.hvd_config.use:
-            self.is_chief = hvd.rank() == 0
-        else:
-            self.is_chief = True
+        self.is_chief = context.distributed.rank == 0
 
-        if self.hvd_config.use and not self.is_chief:
+        if context.distributed.backend == "horovod" and not self.is_chief:
```
Comment (rb-determined-ai) [inline harness/determined/_trial_controller.py:41]: prolly this logging setup is a good default regardless of the backend.

File: harness/determined/estimator/_estimator_context.py [hunk 5]
```diff
@@ -60,7 +68,7 @@ def wrap_optimizer(self, optimizer: Any) -> Any:
             return optimizer
 
         self.optimizer_initialized = True
-        if not self.hvd_config.use:
+        if not self.distributed.backend == "horovod":
```
Comment (rb-determined-ai) [inline harness/determined/estimator/_estimator_context.py:71]: does `distributed.backend == "horovod"` indicate:
1. that we are using horovod for dtrain, and we are currently running under horovodrun, or
2. that we are using horovod for dtrain, even though we may be on a single gpu at the moment (no horovodrun)?

Also, what happens if the backend is something other than horovod, but we are in distributed training?  I'm in a rush with this review, but do you assert that EstimatorTrial only works with horovod somewhere?

**Review by rb-determined-ai — COMMENTED** [review]
it's friday afternoon. I'm only halfway through this review, but that's all I got for this week, sorry.

Removing horovod context from everywhere looks great, I love it!
